File: adapter/py/5_sensor_adapter.py
"""
Adapter Pattern - 感測器統一介面範例

情境：
- 工廠有多個不同品牌的 PLC (可程式邏輯控制器)
- 每個品牌的讀取介面都不同
- 需要統一介面來讀取所有感測器的資料

品牌：
- Siemens (西門子)
- Mitsubishi (三菱)
- Omron (歐姆龍)
"""

from abc import ABC, abstractmethod
from typing import Dict, List, Optional
from datetime import datetime
import random


# ============================================================
# 第一部分：不同品牌的 PLC API (第三方硬體 SDK)
# ============================================================

class SiemensPLC:
    """西門子 PLC SDK"""

    def __init__(self, ip: str = "192.168.1.100", rack: int = 0, slot: int = 1):
        self._ip = ip
        self._rack = rack
        self._slot = slot

    def read_db(self, db_number: int, start: int, length: int) -> bytes:
        """
        讀取資料區塊 (Data Block)
        回傳 raw bytes，需要自行解析
        """
        # 模擬讀取 (實際會透過 S7 協定讀取)
        return b'\x00\x64\x00\x00\x00\x00'  # 模擬溫度: 100

    def read_multiple(self, addresses: List[tuple]) -> Dict[str, bytes]:
        """
        讀取多個位址
        地址格式: (db_number, start_byte, data_type)
        """
        result = {}
        for addr in addresses:
            db, start, length = addr
            result[f"DB{db}.DBX{start}"] = self.read_db(db, start, length)
        return result


class MitsubishiPLC:
    """三菱 PLC SDK"""

    def __init__(self, station: int = 0):
        self._station = station

    def read_device(self, device: str, count: int = 1) -> List[int]:
        """
        讀取裝置 (如 D, M, X, Y)
        device: "D100" 表示資料暫存器 100
        回傳 int 陣列
        """
        # 模擬讀取
        return [random.randint(20, 30) for _ in range(count)]

    def read_devices_batch(self, devices: List[str]) -> Dict[str, int]:
        """批次讀取多個裝置"""
        result = {}
        for device in devices:
            value = self.read_device(device)[0]
            result[device] = value
        return result


class OmronPLC:
    """歐姆龍 PLC SDK"""

    def __init__(self, ip: str = "192.168.1.200", port: int = 9600):
        self._ip = ip
        self._port = port

    def read_cio(self, channel: int, word: int) -> int:
        """
        讀取 CIO 區域
        channel: 通道 (0-999)
        word: 字元位置
        """
        # 模擬讀取
        return random.randint(100, 150)

    def read_range(self, area: str, start: int, count: int) -> List[int]:
        """
        讀取某區域範圍
        area: "DM", "CIO", "WR", "HR"
        """
        # 模擬讀取
        return [random.randint(20, 35) for _ in range(count)]


# ============================================================
# 第二部分：統一感測器介面 (Target Interface)
# ============================================================

class SensorReader(ABC):
    """統一感測器讀取介面"""

    @abstractmethod
    def get_temperature(self, sensor_id: str) -> Dict:
        """
        讀取溫度感測器
        Returns:
            {
                "sensor_id": "string",
                "value": float,
                "unit": "°C",
                "timestamp": "ISO8601",
                "status": "ok|error"
            }
        """
        pass

    @abstractmethod
    def get_humidity(self, sensor_id: str) -> Dict:
        """讀取濕度感測器"""
        pass

    @abstractmethod
    def get_all_sensors(self) -> List[Dict]:
        """讀取所有感測器"""
        pass


# ============================================================
# 第三部分：轉接器實作
# ============================================================

class SiemensSensorAdapter(SensorReader):
    """西門子感測器轉接器"""

    # PLC 記憶體位址對應表
    ADDRESS_MAP = {
        "temp_1": (1, 0, 2),      # DB1.DBD0 - 溫度感測器 1
        "temp_2": (1, 2, 2),      # DB1.DBD2 - 溫度感測器 2
        "humidity_1": (1, 4, 2),  # DB1.DBD4 - 濕度感測器 1
    }

    def __init__(self, plc: SiemensPLC):
        self._plc = plc

    def _read_raw(self, address_key: str) -> int:
        """從 PLC 讀取原始資料"""
        addr = self.ADDRESS_MAP.get(address_key, (1, 0, 2))
        raw = self._plc.read_db(addr[0], addr[1], addr[2])
        # 解析 bytes (假設是 big-endian, 2 bytes)
        return int.from_bytes(raw[:2], 'big', signed=True)

    def get_temperature(self, sensor_id: str) -> Dict:
        value = self._read_raw(f"temp_{sensor_id.split('_')[-1]}")
        return {
            "sensor_id": f"siemens_temp_{sensor_id}",
            "value": float(value) / 10.0,  # 除以 10 得到實際溫度
            "unit": "°C",
            "timestamp": datetime.now().isoformat(),
            "status": "ok"
        }

    def get_humidity(self, sensor_id: str) -> Dict:
        value = self._read_raw(f"humidity_{sensor_id.split('_')[-1]}")
        return {
            "sensor_id": f"siemens_humidity_{sensor_id}",
            "value": float(value) / 10.0,
            "unit": "%RH",
            "timestamp": datetime.now().isoformat(),
            "status": "ok"
        }

    def get_all_sensors(self) -> List[Dict]:
        sensors = []
        for i in range(1, 3):
            sensors.append(self.get_temperature(f"temp_{i}"))
            sensors.append(self.get_humidity(f"humidity_{i}"))
        return sensors


class MitsubishiSensorAdapter(SensorReader):
    """三菱感測器轉接器"""

    DEVICE_MAP = {
        "temp_1": "D100",   # 資料暫存器 100
        "temp_2": "D101",   # 資料暫存器 101
        "humidity_1": "D102",  # 資料暫存器 102
    }

    def __init__(self, plc: MitsubishiPLC):
        self._plc = plc

    def get_temperature(self, sensor_id: str) -> Dict:
        device = self.DEVICE_MAP.get(f"temp_{sensor_id.split('_')[-1]}", "D100")
        value = self._plc.read_device(device)[0]
        return {
            "sensor_id": f"mitsubishi_temp_{sensor_id}",
            "value": float(value),
            "unit": "°C",
            "timestamp": datetime.now().isoformat(),
            "status": "ok"
        }

    def get_humidity(self, sensor_id: str) -> Dict:
        device = self.DEVICE_MAP.get(f"humidity_{sensor_id.split('_')[-1]}", "D102")
        value = self._plc.read_device(device)[0]
        return {
            "sensor_id": f"mitsubishi_humidity_{sensor_id}",
            "value": float(value),
            "unit": "%RH",
            "timestamp": datetime.now().isoformat(),
            "status": "ok"
        }

    def get_all_sensors(self) -> List[Dict]:
        devices = ["D100", "D101", "D102"]
        values = self._plc.read_devices_batch(devices)

        sensors = [
            {"sensor_id": "mitsubishi_temp_temp_1", "value": values["D100"], "unit": "°C", "timestamp": datetime.now().isoformat(), "status": "ok"},
            {"sensor_id": "mitsubishi_temp_temp_2", "value": values["D101"], "unit": "°C", "timestamp": datetime.now().isoformat(), "status": "ok"},
            {"sensor_id": "mitsubishi_humidity_humidity_1", "value": values["D102"], "unit": "%RH", "timestamp": datetime.now().isoformat(), "status": "ok"},
        ]
        return sensors


class OmronSensorAdapter(SensorReader):
    """歐姆龍感測器轉接器"""

    AREA_MAP = {
        "temp_1": ("DM", 100),   # DM 區域 100
        "temp_2": ("DM", 101),   # DM 區域 101
        "humidity_1": ("DM", 102),  # DM 區域 102
    }

    def __init__(self, plc: OmronPLC):
        self._plc = plc

    def get_temperature(self, sensor_id: str) -> Dict:
        area, word = self.AREA_MAP.get(f"temp_{sensor_id.split('_')[-1]}", ("DM", 100))
        value = self._plc.read_cio(0, word)
        return {
            "sensor_id": f"omron_temp_{sensor_id}",
            "value": float(value) / 10.0,
            "unit": "°C",
            "timestamp": datetime.now().isoformat(),
            "status": "ok"
        }

    def get_humidity(self, sensor_id: str) -> Dict:
        area, word = self.AREA_MAP.get(f"humidity_{sensor_id.split('_')[-1]}", ("DM", 102))
        value = self._plc.read_cio(0, word)
        return {
            "sensor_id": f"omron_humidity_{sensor_id}",
            "value": float(value) / 10.0,
            "unit": "%RH",
            "timestamp": datetime.now().isoformat(),
            "status": "ok"
        }

    def get_all_sensors(self) -> List[Dict]:
        values = self._plc.read_range("DM", 100, 3)
        sensors = [
            {"sensor_id": "omron_temp_temp_1", "value": values[0]/10.0, "unit": "°C", "timestamp": datetime.now().isoformat(), "status": "ok"},
            {"sensor_id": "omron_temp_temp_2", "value": values[1]/10.0, "unit": "°C", "timestamp": datetime.now().isoformat(), "status": "ok"},
            {"sensor_id": "omron_humidity_humidity_1", "value": values[2]/10.0, "unit": "%RH", "timestamp": datetime.now().isoformat(), "status": "ok"},
        ]
        return sensors


# ============================================================
# 第四部分：感測器集中管理 (配合其他模式)
# ============================================================

class SensorManager:
    """感測器管理器 - 統一管理所有品牌的感測器"""

    def __init__(self):
        self._adapters: Dict[str, SensorReader] = {}

    def register_adapter(self, brand: str, adapter: SensorReader) -> None:
        """註冊某品牌的轉接器"""
        self._adapters[brand] = adapter

    def read_sensor(self, brand: str, sensor_type: str, sensor_id: str) -> Optional[Dict]:
        """讀取指定品牌的感測器"""
        adapter = self._adapters.get(brand)
        if not adapter:
            return None

        if sensor_type == "temperature":
            return adapter.get_temperature(sensor_id)
        elif sensor_type == "humidity":
            return adapter.get_humidity(sensor_id)
        return None

    def read_all(self) -> List[Dict]:
        """讀取所有感測器"""
        all_sensors = []
        for brand, adapter in self._adapters.items():
            sensors = adapter.get_all_sensors()
            for sensor in sensors:
                sensor["brand"] = brand
            all_sensors.extend(sensors)
        return all_sensors


# ============================================================
# 第五部分：執行範例
# ============================================================

def display_sensor_data(sensors: List[Dict]) -> None:
    """顯示感測器資料"""

    print("\n" + "=" * 80)
    print(f"📊 工廠感測器統一讀取結果 ({len(sensors)} 筆資料)")
    print("=" * 80)

    print(f"\n{'品牌':<12} {'感測器ID':<28} {'數值':<10} {'單位':<6}")
    print("-" * 60)

    for sensor in sensors:
        print(f"{sensor.get('brand', 'N/A'):<12} {sensor['sensor_id']:<28} "
              f"{sensor['value']:<10.2f} {sensor['unit']:<6}")

    print()


def demonstrate_single_brand(adapter: SensorReader, brand: str) -> None:
    """示範單一品牌的讀取"""

    print(f"\n🔌 {brand} 感測器讀取示範:")
    print("-" * 40)

    # 讀取溫度
    temp = adapter.get_temperature("1")
    print(f"  🌡️  溫度: {temp['value']:.1f} {temp['unit']}")

    # 讀取濕度
    humidity = adapter.get_humidity("1")
    print(f"  💧 濕度: {humidity['value']:.1f} {humidity['unit']}")


if __name__ == "__main__":
    print("=" * 80)
    print("Adapter Pattern - 感測器統一介面範例")
    print("=" * 80)

    # 建立各品牌的 PLC 實例
    siemens_plc = SiemensPLC()
    mitsubishi_plc = MitsubishiPLC()
    omron_plc = OmronPLC()

    # 建立轉接器
    siemens_adapter = SiemensSensorAdapter(siemens_plc)
    mitsubishi_adapter = MitsubishiSensorAdapter(mitsubishi_plc)
    omron_adapter = OmronSensorAdapter(omron_plc)

    # 示範單一品牌
    demonstrate_single_brand(siemens_adapter, "Siemens")
    demonstrate_single_brand(mitsubishi_adapter, "Mitsubishi")
    demonstrate_single_brand(omron_adapter, "Omron")

    # 建立感測器管理器
    manager = SensorManager()
    manager.register_adapter("siemens", siemens_adapter)
    manager.register_adapter("mitsubishi", mitsubishi_adapter)
    manager.register_adapter("omron", omron_adapter)

    # 統一讀取所有感測器
    all_sensors = manager.read_all()
    display_sensor_data(all_sensors)

    print("\n🎯 重點說明：")
    print("1. 每個品牌的 PLC API 都不同 (Siemens, Mitsubishi, Omron)")
    print("2. 透過轉接器，每個品牌都能提供統一的 SensorReader 介面")
    print("3. 監控系統只需要知道 SensorReader 介面，不需要知道底層 PLC 差異")
    print("4. 未來新增品牌，只需實作新的 Adapter")
    print("5. 這個範例展示了 IoT 與工業自動化的常見應用場景")
    print("=" * 80)

File: adapter/cs/5_SensorAdapter.cs
namespace AdapterExamples
{
    // ============================================================
    // 第五部分：感測器統一介面範例
    // ============================================================

    #region 不同品牌的 PLC SDK (第三方硬體 SDK)

    /// <summary>
    /// 西門子 PLC SDK
    /// </summary>
    public class SiemensPlc
    {
        private readonly string _ip;
        private readonly int _rack;
        private readonly int _slot;

        public SiemensPlc(string ip = "192.168.1.100", int rack = 0, int slot = 1)
        {
            _ip = ip;
            _rack = rack;
            _slot = slot;
        }

        /// <summary>
        /// 讀取資料區塊 (Data Block)
        /// </summary>
        public byte[] ReadDb(int dbNumber, int start, int length)
        {
            // 模擬讀取 (實際會透過 S7 協定讀取)
            // 模擬溫度: 100 → 實際要除以 10
            byte[] data = new byte[length];
            Random.Shared.NextBytes(data);
            data[0] = 0x00;
            data[1] = 0x64; // 100
            return data;
        }
    }

    /// <summary>
    /// 三菱 PLC SDK
    /// </summary>
    public class MitsubishiPlc
    {
        private readonly int _station;

        public MitsubishiPlc(int station = 0)
        {
            _station = station;
        }

        /// <summary>
        /// 讀取裝置 (如 D, M, X, Y)
        /// </summary>
        public int[] ReadDevice(string device, int count = 1)
        {
            // 模擬讀取
            return Enumerable.Range(0, count)
                .Select(_ => Random.Shared.Next(20, 30))
                .ToArray();
        }
    }

    /// <summary>
    /// 歐姆龍 PLC SDK
    /// </summary>
    public class OmronPlc
    {
        private readonly string _ip;
        private readonly int _port;

        public OmronPlc(string ip = "192.168.1.200", int port = 9600)
        {
            _ip = ip;
            _port = port;
        }

        /// <summary>
        /// 讀取 CIO 區域
        /// </summary>
        public int ReadCio(int channel, int word)
        {
            // 模擬讀取
            return Random.Shared.Next(100, 150);
        }
    }

    #endregion

    #region 統一感測器介面

    /// <summary>
    /// 統一感測器讀取介面
    /// </summary>
    public interface ISensorReader
    {
        /// <summary>
        /// 讀取溫度感測器
        /// </summary>
        SensorData GetTemperature(string sensorId);

        /// <summary>
        /// 讀取濕度感測器
        /// </summary>
        SensorData GetHumidity(string sensorId);

        /// <summary>
        /// 讀取所有感測器
        /// </summary>
        List<SensorData> GetAllSensors();
    }

    /// <summary>
    /// 感測器資料
    /// </summary>
    public record SensorData(
        string SensorId,
        double Value,
        string Unit,
        string Timestamp,
        string Status
    );

    #endregion

    #region 轉接器實作

    /// <summary>
    /// 西門子感測器轉接器
    /// </summary>
    public class SiemensSensorAdapter : ISensorReader
    {
        private readonly SiemensPlc _plc;

        // PLC 記憶體位址對應表
        private readonly Dictionary<string, (int db, int start, int length)> _addressMap = new()
        {
            { "temp_1", (1, 0, 2) },      // DB1.DBD0
            { "temp_2", (1, 2, 2) },      // DB1.DBD2
            { "humidity_1", (1, 4, 2) },  // DB1.DBD4
        };

        public SiemensSensorAdapter(SiemensPlc plc)
        {
            _plc = plc;
        }

        private int ReadRaw(string addressKey)
        {
            var addr = _addressMap.GetValueOrDefault(addressKey, (1, 0, 2));
            var raw = _plc.ReadDb(addr.db, addr.start, addr.length);
            return BitConverter.ToInt16(raw[..2], 0);
        }

        public SensorData GetTemperature(string sensorId)
        {
            string key = $"temp_{sensorId.Split('_').Last()}";
            int rawValue = ReadRaw(key);
            return new SensorData(
                SensorId: $"siemens_temp_{sensorId}",
                Value: rawValue / 10.0,
                Unit: "°C",
                Timestamp: DateTime.UtcNow.ToString("o"),
                Status: "ok"
            );
        }

        public SensorData GetHumidity(string sensorId)
        {
            string key = $"humidity_{sensorId.Split('_').Last()}";
            int rawValue = ReadRaw(key);
            return new SensorData(
                SensorId: $"siemens_humidity_{sensorId}",
                Value: rawValue / 10.0,
                Unit: "%RH",
                Timestamp: DateTime.UtcNow.ToString("o"),
                Status: "ok"
            );
        }

        public List<SensorData> GetAllSensors()
        {
            var sensors = new List<SensorData>();
            foreach (var i in new[] { "1", "2" })
            {
                sensors.Add(GetTemperature($"temp_{i}"));
                sensors.Add(GetHumidity($"humidity_{i}"));
            }
            return sensors;
        }
    }

    /// <summary>
    /// 三菱感測器轉接器
    /// </summary>
    public class MitsubishiSensorAdapter : ISensorReader
    {
        private readonly MitsubishiPlc _plc;

        private readonly Dictionary<string, string> _deviceMap = new()
        {
            { "temp_1", "D100" },
            { "temp_2", "D101" },
            { "humidity_1", "D102" },
        };

        public MitsubishiSensorAdapter(MitsubishiPlc plc)
        {
            _plc = plc;
        }

        public SensorData GetTemperature(string sensorId)
        {
            string key = $"temp_{sensorId.Split('_').Last()}";
            string device = _deviceMap.GetValueOrDefault(key, "D100");
            int value = _plc.ReadDevice(device)[0];
            return new SensorData(
                SensorId: $"mitsubishi_temp_{sensorId}",
                Value: value,
                Unit: "°C",
                Timestamp: DateTime.UtcNow.ToString("o"),
                Status: "ok"
            );
        }

        public SensorData GetHumidity(string sensorId)
        {
            string key = $"humidity_{sensorId.Split('_').Last()}";
            string device = _deviceMap.GetValueOrDefault(key, "D102");
            int value = _plc.ReadDevice(device)[0];
            return new SensorData(
                SensorId: $"mitsubishi_humidity_{sensorId}",
                Value: value,
                Unit: "%RH",
                Timestamp: DateTime.UtcNow.ToString("o"),
                Status: "ok"
            );
        }

        public List<SensorData> GetAllSensors()
        {
            var devices = new[] { "D100", "D101", "D102" };
            var values = devices.ToDictionary(d => d, d => _plc.ReadDevice(d)[0]);

            return new List<SensorData>
            {
                new("mitsubishi_temp_temp_1", values["D100"], "°C", DateTime.UtcNow.ToString("o"), "ok"),
                new("mitsubishi_temp_temp_2", values["D101"], "°C", DateTime.UtcNow.ToString("o"), "ok"),
                new("mitsubishi_humidity_humidity_1", values["D102"], "%RH", DateTime.UtcNow.ToString("o"), "ok"),
            };
        }
    }

    /// <summary>
    /// 歐姆龍感測器轉接器
    /// </summary>
    public class OmronSensorAdapter : ISensorReader
    {
        private readonly OmronPlc _plc;

        private readonly Dictionary<string, (string area, int word)> _areaMap = new()
        {
            { "temp_1", ("DM", 100) },
            { "temp_2", ("DM", 101) },
            { "humidity_1", ("DM", 102) },
        };

        public OmronSensorAdapter(OmronPlc plc)
        {
            _plc = plc;
        }

        public SensorData GetTemperature(string sensorId)
        {
            string key = $"temp_{sensorId.Split('_').Last()}";
            var (area, word) = _areaMap.GetValueOrDefault(key, ("DM", 100));
            int value = _plc.ReadCio(0, word);
            return new SensorData(
                SensorId: $"omron_temp_{sensorId}",
                Value: value / 10.0,
                Unit: "°C",
                Timestamp: DateTime.UtcNow.ToString("o"),
                Status: "ok"
            );
        }

        public SensorData GetHumidity(string sensorId)
        {
            string key = $"humidity_{sensorId.Split('_').Last()}";
            var (area, word) = _areaMap.GetValueOrDefault(key, ("DM", 102));
            int value = _plc.ReadCio(0, word);
            return new SensorData(
                SensorId: $"omron_humidity_{sensorId}",
                Value: value / 10.0,
                Unit: "%RH",
                Timestamp: DateTime.UtcNow.ToString("o"),
                Status: "ok"
            );
        }

        public List<SensorData> GetAllSensors()
        {
            var values = new[] { 110, 115, 120 }; // 模擬讀取值
            return new List<SensorData>
            {
                new("omron_temp_temp_1", values[0] / 10.0, "°C", DateTime.UtcNow.ToString("o"), "ok"),
                new("omron_temp_temp_2", values[1] / 10.0, "°C", DateTime.UtcNow.ToString("o"), "ok"),
                new("omron_humidity_humidity_1", values[2] / 10.0, "%RH", DateTime.UtcNow.ToString("o"), "ok"),
            };
        }
    }

    #endregion

    #region 感測器管理器

    /// <summary>
    /// 感測器管理器 - 統一管理所有品牌的感測器
    /// </summary>
    public class SensorManager
    {
        private readonly Dictionary<string, ISensorReader> _adapters = new();

        public void RegisterAdapter(string brand, ISensorReader adapter)
        {
            _adapters[brand] = adapter;
        }

        public SensorData? ReadSensor(string brand, string sensorType, string sensorId)
        {
            if (!_adapters.TryGetValue(brand, out var adapter))
                return null;

            return sensorType switch
            {
                "temperature" => adapter.GetTemperature(sensorId),
                "humidity" => adapter.GetHumidity(sensorId),
                _ => null
            };
        }

        public List<SensorData> ReadAll()
        {
            var allSensors = new List<SensorData>();
            foreach (var (_, adapter) in _adapters)
            {
                var sensors = adapter.GetAllSensors();
                allSensors.AddRange(sensors);
            }
            return allSensors;
        }
    }

    #endregion

    /// <summary>
    /// 感測器範例執行
    /// </summary>
    public static class SensorAdapter
    {
        public static void Run()
        {
            Console.WriteLine("\n" + "=".PadRight(60, '='));
            Console.WriteLine("5. 感測器統一介面範例");
            Console.WriteLine("=".PadRight(60, '='));

            // 建立各品牌的 PLC 實例
            var siemensPlc = new SiemensPlc();
            var mitsubishiPlc = new MitsubishiPlc();
            var omronPlc = new OmronPlc();

            // 建立轉接器
            var siemensAdapter = new SiemensSensorAdapter(siemensPlc);
            var mitsubishiAdapter = new MitsubishiSensorAdapter(mitsubishiPlc);
            var omronAdapter = new OmronSensorAdapter(omronPlc);

            // 示範單一品牌讀取
            Console.WriteLine("\n🔌 各品牌感測器讀取示範:");
            Console.WriteLine("-".PadRight(50, '-'));

            void ShowBrand(string brand, ISensorReader adapter)
            {
                Console.WriteLine($"\n  {brand}:");
                var temp = adapter.GetTemperature("1");
                Console.WriteLine($"    🌡️  溫度: {temp.Value:F1} {temp.Unit}");
                var humidity = adapter.GetHumidity("1");
                Console.WriteLine($"    💧 濕度: {humidity.Value:F1} {humidity.Unit}");
            }

            ShowBrand("Siemens", siemensAdapter);
            ShowBrand("Mitsubishi", mitsubishiAdapter);
            ShowBrand("Omron", omronAdapter);

            // 建立感測器管理器
            var manager = new SensorManager();
            manager.RegisterAdapter("siemens", siemensAdapter);
            manager.RegisterAdapter("mitsubishi", mitsubishiAdapter);
            manager.RegisterAdapter("omron", omronAdapter);

            // 統一讀取所有感測器
            var allSensors = manager.ReadAll();

            Console.WriteLine($"\n📊 工廠感測器統一讀取結果 ({allSensors.Count} 筆資料)");
            Console.WriteLine("-".PadRight(60, '-'));
            Console.WriteLine($"{"品牌",-12} {"感測器ID",-28} {"數值",-8} {"單位",-6}");
            Console.WriteLine("-".PadRight(60, '-'));

            // 顯示感測器資料 (這裡用簡化的方式)
            foreach (var adapter in new[] { siemensAdapter, mitsubishiAdapter, omronAdapter })
            {
                var brand = adapter.GetType().Name.Replace("SensorAdapter", "");
                foreach (var sensor in adapter.GetAllSensors())
                {
                    Console.WriteLine($"{brand,-12} {sensor.SensorId,-28} {sensor.Value,-8:F1} {sensor.Unit,-6}");
                }
            }

            Console.WriteLine("\n🎯 重點說明：");
            Console.WriteLine("   1. 每個品牌的 PLC API 都不同 (Siemens, Mitsubishi, Omron)");
            Console.WriteLine("   2. 透過轉接器，每個品牌都能提供統一的 ISensorReader 介面");
            Console.WriteLine("   3. 監控系統只需要知道 ISensorReader 介面");
            Console.WriteLine("   4. 未來新增品牌，只需實作新的 Adapter");
            Console.WriteLine("   5. 這個範例展示了 IoT 與工業自動化的常見應用場景");
        }
    }
}